File: python/processing/subtitle_format.py
"""Pure lyric line wrapping and rolling-SRT formatting."""

import re


def format_srt_timestamp(seconds):
    milliseconds = max(0, round(seconds * 1000))
    hours, milliseconds = divmod(milliseconds, 3_600_000)
    minutes, milliseconds = divmod(milliseconds, 60_000)
    seconds, milliseconds = divmod(milliseconds, 1_000)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d},{milliseconds:03d}"


def normalized_text(text):
    return "".join(character.casefold() for character in text if character.isalnum())


def read_srt_segments(filename):
    """Read creator captions as alignment-ready segments, removing music markup."""
    with open(filename, encoding="utf-8-sig") as source:
        content = source.read().replace("\r\n", "\n")

    segments = []
    for block in re.split(r"\n{2,}", content.strip()):
        lines = block.splitlines()
        timing_index = next((index for index, line in enumerate(lines) if "-->" in line), None)
        if timing_index is None:
            continue
        match = re.match(r"\s*(\d{2}):(\d{2}):(\d{2})[,.](\d{3})\s*-->\s*(\d{2}):(\d{2}):(\d{2})[,.](\d{3})", lines[timing_index])
        if not match:
            continue
        values = [int(value) for value in match.groups()]
        start = values[0] * 3600 + values[1] * 60 + values[2] + values[3] / 1000
        end = values[4] * 3600 + values[5] * 60 + values[6] + values[7] / 1000
        text = " ".join(lines[timing_index + 1:])
        text = re.sub(r"<[^>]+>", " ", text).replace("♪", " ")
        text = re.sub(r"\[[^\]]+\]", " ", text)
        text = re.sub(r"\s+", " ", text).strip()
        if end > start and normalized_text(text):
            segments.append({"start": start, "end": end, "text": text})
    return segments


def preserve_unaligned_text(raw_segments, aligned_segments):
    """Restore a raw ASR segment if forced alignment dropped any of its text."""
    buckets = [[] for _ in raw_segments]
    raw_index = 0
    for aligned in aligned_segments:
        start = aligned.get("start")
        end = aligned.get("end")
        if start is None or end is None:
            continue
        for index in range(raw_index, len(raw_segments)):
            raw = raw_segments[index]
            if start > raw["end"] + 0.05:
                raw_index = index + 1
                continue
            if start >= raw["start"] - 0.05 and end <= raw["end"] + 0.05:
                buckets[index].append(aligned)
                raw_index = index
            break

    preserved = []
    for raw, aligned in zip(raw_segments, buckets):
        raw_coverage = normalized_text(raw.get("text", ""))
        aligned_coverage = normalized_text("".join(segment.get("text", "") for segment in aligned))
        if raw_coverage and raw_coverage != aligned_coverage:
            preserved.append({
                "text": raw["text"],
                "start": raw["start"],
                "end": raw["end"],
                "words": [],
            })
        else:
            preserved.extend(aligned)
    return preserved


def timed_words(result):
    """Return aligned words, estimating a segment only if it has no word timings."""
    words = []
    no_spaces = result["language"] in {"zh", "yue", "ja", "ko", "th"}
    for segment in result["segments"]:
        segment_words = segment.get("words") or []
        if not segment_words:
            text = segment.get("text", "").strip()
            units = list(text) if no_spaces else text.split()
            segment_words = [{"word": word} for word in units]
        visible_words = [word for word in segment_words if word.get("word", "").strip()]
        if not visible_words:
            continue
        duration = max(0, segment["end"] - segment["start"])
        for index, word in enumerate(visible_words):
            start = word.get("start")
            end = word.get("end")
            if start is None or end is None:
                start = segment["start"] + duration * index / len(visible_words)
                end = segment["start"] + duration * (index + 1) / len(visible_words)
            words.append({"text": word["word"].strip(), "start": start, "end": end})
    return words


def lyric_lines(result, max_width, min_width):
    """Wrap aligned words into readable lyric lines without losing timing."""
    def is_sentence_end(text):
        return text.rstrip('”"\'»)]}').endswith((".", "。"))

    def is_phrase_end(text):
        return text.rstrip('”"\'»)]}').endswith((",", "，", "、", "!", "?", "！", "？"))

    def add_line():
        lines.append({
            "text": "".join(item["text"] for item in current) if no_spaces else " ".join(item["text"] for item in current),
            "start": current[0]["start"],
            "end": current[-1]["end"],
        })

    no_spaces = result["language"] in {"zh", "yue", "ja", "ko", "th"}
    lines = []
    current = []
    current_width = 0
    carry_short_comma = False
    for word in timed_words(result):
        separator_width = 0 if no_spaces or not current else 1
        word_width = len(word["text"])
        pause_before_word = word["start"] - current[-1]["end"] if current else 0
        natural_phrase_break = (
            no_spaces
            and current_width >= max_width / 2
            and pause_before_word >= 0.45
            and not carry_short_comma
        )
        if current and (natural_phrase_break or current_width + separator_width + word_width > max_width):
            add_line()
            current = []
            current_width = 0
            separator_width = 0
        current.append(word)
        current_width += separator_width + word_width
        carry_short_comma = False
        if is_sentence_end(word["text"]):
            add_line()
            current = []
            current_width = 0
        elif no_spaces and is_phrase_end(word["text"]):
            if current_width >= min_width:
                add_line()
                current = []
                current_width = 0
            else:
                carry_short_comma = True
    if current:
        add_line()
    return lines


def rolling_cues(result, max_width, min_width):
    """Build two independently advancing lyric rows.

    Yields (start, end, text, active_row) — active_row is the 0-based index of
    the currently-singing line within `text.split("\\n")`. Encoding this
    explicitly avoids the client having to diff consecutive cues to guess which
    row is active, which is unreliable when two rows share identical text.
    """
    lines = lyric_lines(result, max_width, min_width)
    groups = []
    for line in lines:
        if not groups or line["start"] - groups[-1][-1]["end"] > 2:
            groups.append([line])
        else:
            groups[-1].append(line)

    cues = []
    for group in groups:
        top = group[0]
        bottom = group[1] if len(group) > 1 else None
        next_line = 2
        current_time = top["start"]
        while top or bottom:
            ending_top = top is not None and (bottom is None or top["end"] <= bottom["end"])
            ending = top if ending_top else bottom
            if ending["end"] > current_time:
                displayed = [line["text"] for line in (top, bottom) if line is not None]
                # active_row is the index into `displayed`, not into (top, bottom).
                # If only bottom remains, it slides up to index 0.
                active_row = 0 if ending_top or top is None else 1
                cues.append((current_time, ending["end"], "\n".join(displayed), active_row))
            current_time = ending["end"]
            replacement = group[next_line] if next_line < len(group) else None
            next_line += 1
            if ending_top:
                top = replacement
            else:
                bottom = replacement
    return cues


def write_rolling_srt(result, output_path, max_width, min_width):
    with open(output_path, "w", encoding="utf-8") as output:
        for index, (start, end, text, active_row) in enumerate(rolling_cues(result, max_width, min_width), start=1):
            # `A<row>` is appended after the timing arrow. Standard SRT parsers
            # take only the two timestamps from this line and ignore trailing
            # tokens, so external players remain compatible.
            output.write(
                f"{index}\n{format_srt_timestamp(start)} --> {format_srt_timestamp(end)} A{active_row}\n{text}\n\n"
            )

File: shared/subtitleFormat.ts
export interface TranscriptWord {
  end: number
  start: number
  text: string
  type: 'audio_event' | 'spacing' | 'word'
}

interface TimedWord {
  end: number
  start: number
  text: string
}

interface LyricWord extends TimedWord {
  characterEnd: number
  characterStart: number
}

interface LyricLine extends TimedWord {
  words: LyricWord[]
}

export interface RollingCue extends TimedWord {
  activeEnd: number
  activeRow: number
  activeStart: number
  activeWordEnd: number
  activeWordStart: number
}

const NO_SPACE_LANGUAGES = new Set([
  'ja', 'jpn', 'ko', 'kor', 'lo', 'lao', 'my', 'mya', 'th', 'tha', 'yue', 'zh', 'zho',
])

export function createRollingSrt (
  words: TranscriptWord[],
  language: string,
  maxWidth: number,
  minWidth: number,
): string {
  return rollingCues(words, language, maxWidth, minWidth)
    .map((cue, index) => [
      index + 1,
      `${formatSrtTimestamp(cue.start)} --> ${formatSrtTimestamp(cue.end)} A${cue.activeRow} C${cue.activeStart}-${cue.activeEnd} P${Math.round(cue.activeWordStart * 1000)}-${Math.round(cue.activeWordEnd * 1000)}`,
      cue.text,
      '',
    ].join('\n'))
    .join('\n')
}

export function rollingCues (
  words: TranscriptWord[],
  language: string,
  maxWidth: number,
  minWidth: number,
): RollingCue[] {
  const lines = lyricLines(words, language, maxWidth, minWidth)
  const groups: LyricLine[][] = []
  for (const line of lines) {
    const lastGroup = groups.at(-1)
    if (!lastGroup || line.start - lastGroup.at(-1)!.end > 2) groups.push([line])
    else lastGroup.push(line)
  }

  const cues: RollingCue[] = []
  for (const group of groups) {
    let top: LyricLine | undefined = group[0]
    let bottom: LyricLine | undefined = group[1]
    let nextLine = 2
    let currentTime = top.start
    while (top || bottom) {
      const endingTop = top !== undefined && (bottom === undefined || top.end <= bottom.end)
      const ending = endingTop ? top! : bottom!
      if (ending.end > currentTime) {
        const displayed = [top, bottom].filter((line): line is LyricLine => line !== undefined)
        const activeRow = endingTop || top === undefined ? 0 : 1
        const text = displayed.map(line => line.text).join('\n')
        let wordStart = currentTime
        ending.words.forEach((word, index) => {
          const nextWord = ending.words[index + 1]
          const wordEnd = nextWord ? Math.min(ending.end, Math.max(wordStart, nextWord.start)) : ending.end
          if (wordEnd > wordStart) {
            cues.push({
              activeEnd: word.characterEnd,
              activeRow,
              activeStart: word.characterStart,
              activeWordEnd: word.end,
              activeWordStart: word.start,
              end: wordEnd,
              start: wordStart,
              text,
            })
          }
          wordStart = wordEnd
        })
      }
      currentTime = ending.end
      const replacement = group[nextLine++]
      if (endingTop) top = replacement
      else bottom = replacement
    }
  }
  return cues
}

export function lyricLines (
  words: TranscriptWord[],
  language: string,
  maxWidth: number,
  minWidth: number,
): LyricLine[] {
  const noSpaces = NO_SPACE_LANGUAGES.has(language.toLowerCase())
  const lines: LyricLine[] = []
  let current: TimedWord[] = []
  let currentWidth = 0
  let carryShortComma = false
  const addLine = () => {
    let characterOffset = 0
    const lineWords = current.map((word, index) => {
      if (!noSpaces && index > 0) characterOffset++
      const lineWord = {
        ...word,
        characterEnd: characterOffset + word.text.length,
        characterStart: characterOffset,
      }
      characterOffset = lineWord.characterEnd
      return lineWord
    })
    lines.push({
      end: current.at(-1)!.end,
      start: current[0].start,
      text: current.map(word => word.text).join(noSpaces ? '' : ' '),
      words: lineWords,
    })
  }

  for (const word of timedWords(words)) {
    let separatorWidth = noSpaces || !current.length ? 0 : 1
    const pauseBeforeWord = current.length ? word.start - current.at(-1)!.end : 0
    const naturalPhraseBreak = noSpaces
      && currentWidth >= maxWidth / 2
      && pauseBeforeWord >= 0.45
      && !carryShortComma
    if (current.length && (naturalPhraseBreak || currentWidth + separatorWidth + word.text.length > maxWidth)) {
      addLine()
      current = []
      currentWidth = 0
      separatorWidth = 0
    }
    current.push(word)
    currentWidth += separatorWidth + word.text.length
    carryShortComma = false
    if (isSentenceEnd(word.text)) {
      addLine()
      current = []
      currentWidth = 0
    } else if (noSpaces && isPhraseEnd(word.text)) {
      if (currentWidth >= minWidth) {
        addLine()
        current = []
        currentWidth = 0
      } else carryShortComma = true
    }
  }
  if (current.length) addLine()
  return lines
}

export function formatSrtTimestamp (seconds: number): string {
  let milliseconds = Math.max(0, Math.round(seconds * 1000))
  const hours = Math.floor(milliseconds / 3_600_000)
  milliseconds %= 3_600_000
  const minutes = Math.floor(milliseconds / 60_000)
  milliseconds %= 60_000
  const wholeSeconds = Math.floor(milliseconds / 1000)
  milliseconds %= 1000
  return [hours, minutes, wholeSeconds].map(value => String(value).padStart(2, '0')).join(':')
    + `,${String(milliseconds).padStart(3, '0')}`
}

export interface KaraokeWord {
  end: number
  start: number
  text: string
}

/** Convert edited words into generator transcript input, in time order. */
export function wordsToTranscript (words: KaraokeWord[]): TranscriptWord[] {
  return words
    .filter(word => word.text.trim() && Number.isFinite(word.start) && Number.isFinite(word.end)
      && word.end > word.start)
    .sort((a, b) => a.start - b.start)
    .map(word => ({ end: word.end, start: word.start, text: word.text.trim(), type: 'word' }))
}

/**
 * Extract the timed word list encoded in generated karaoke SRT cues. Cues
 * without the A/C/P word metadata (legacy scripts) are skipped.
 */
export function parseKaraokeWords (source: string): KaraokeWord[] {
  return source.trim().split(/\r?\n\s*\r?\n/).flatMap((block) => {
    const lines = block.split(/\r?\n/)
    const timingIndex = lines.findIndex(line => line.includes('-->'))
    if (timingIndex === -1) return []
    const [, rest] = lines[timingIndex].split('-->')
    const activeMatch = rest.match(/\bA(\d+)\b/)
    const characterMatch = rest.match(/\bC(\d+)-(\d+)\b/)
    const progressMatch = rest.match(/\bP(\d+)-(\d+)\b/)
    if (!activeMatch || !characterMatch || !progressMatch) return []
    const activeLine = lines[timingIndex + 1 + Number(activeMatch[1])]
    if (activeLine === undefined) return []
    const text = activeLine.slice(Number(characterMatch[1]), Number(characterMatch[2])).trim()
    const start = Number(progressMatch[1]) / 1000
    const end = Number(progressMatch[2]) / 1000
    return text && end > start ? [{ text, start, end }] : []
  })
}

function timedWords (words: TranscriptWord[]): TimedWord[] {
  return words.flatMap((word) => {
    const text = word.text.trim()
    return word.type === 'word' && text && Number.isFinite(word.start) && Number.isFinite(word.end)
      && word.end > word.start
      ? [{ text, start: word.start, end: word.end }]
      : []
  })
}

function isSentenceEnd (text: string): boolean {
  return text.replace(/[”"'»\])}]+$/u, '').endsWith('.') || text.replace(/[”"'»\])}]+$/u, '').endsWith('。')
}

function isPhraseEnd (text: string): boolean {
  return /[,，、!?！？][”"'»\])}]*$/u.test(text)
}
